notebooks/hello-universe/Interpreting_CNNs/Interpreting_CNNs.ipynb

```diff
@@ -0,0 +1,803 @@
+   "source": [
+    "**The result is very similar** to our saliency map from backpropagation, with the image being coarser but also showing bot the top and bottom of the galaxy. We can also play around with the selected layer to calculate the output at different steps in the model and see how activation changes throughout the model."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "27909148-c421-48b6-9ecb-a47f56a5dd61",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## 3. RISE Algorithm\n",
+    "The RISE (Randomized Input Sampling for Explanation) Algorithm is another interpretation technique for calculating saliency maps. Instead of calculating gradients from within the model, the RISE implementation works by covering up pieces of the input image, running it through the model, and calculating the average activation in order to determine what parts of the image are most important. As such, this method does not require any access to the inner workings of the model. The algorithm first generates a sequence random of binary grids (called masks), which are placed onto the image. Everything not covered by the mask is removed by multiplying the images together, and the resultant activations are averaged to get our final heatmap. A full description of the algorithm and its variations is provided by <a href='https://arxiv.org/abs/1806.07421'>Petsiuk et al. 2018</a>. \n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1f6c0619-5f75-4f8f-b3c4-2611f987672f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Choose the image to analyze\n",
+    "img_idx = 6\n",
+    "\n",
+    "# We can change the index to any number in range of the test set\n",
+    "image = X_test[img_idx]\n",
+    "\n",
+    "N = 1000  # Number of masks\n",
+    "s = 8     # Size of the grid\n",
+    "p1 = 0.5  # Probability of the cell being set to 1\n",
+    "\n",
+    "cell_size = np.ceil(np.array(input_shape[:2]) / s).astype(int)\n",
+    "up_size = (s * cell_size).astype(int)\n",
+    "\n",
+    "grid = np.random.rand(N, s, s) < p1\n",
+    "masks = np.empty((N, *input_shape[:2]))\n",
+    "\n",
+    "for i in range(N):\n",
+    "    # Randomly place the grid on the image\n",
+    "    x = np.random.randint(0, input_shape[0]-s)\n",
+    "    y = np.random.randint(0, input_shape[1]-s)\n",
+    "    mask = np.pad(grid[i], ((x, input_shape[0]-x-s), (y, input_shape[0]-y-s)), 'constant', constant_values=(0, 0))\n",
+    "    mask = mask[:input_shape[0], :input_shape[1]]\n",
+    "    masks[i] = mask\n",
+    "\n",
+    "masks = masks.reshape(-1, *input_shape[:2], 1)\n",
+    "\n",
+    "N = len(masks)\n",
+    "pred_masks = cnn.predict(image * masks)\n",
+    "pred_masks = np.expand_dims(pred_masks, axis=-1)\n",
+    "pred_masks = np.expand_dims(pred_masks, axis=-1) # Reshape pred_masks for broadcasting\n",
+    "heatmap = (pred_masks * masks).sum(axis=0)\n",
+    "heatmap = heatmap / N / p1"
```

Comment:
Is this baked into to Keras or tensorflow somewhere?  While the previous two are straightforward, this seems complex enough it could benefit from an implementation in a library.  It still makes sense to leave this here to help the reader understand, but if there's a way to make it using a library implementation perhaps that should be mentioned?

Reply:
(To be clear I'm just guessing there's some way to do it from a library.  If not, just ignore this!)